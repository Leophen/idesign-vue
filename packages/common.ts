import _ from 'lodash'
import {
  Component,
  getCurrentInstance,
  Slots,
  VNode,
} from 'vue'

// 通用位置类型
export type placementType =
  | 'top'
  | 'top-left'
  | 'top-right'
  | 'bottom'
  | 'bottom-left'
  | 'bottom-right'
  | 'left'
  | 'left-top'
  | 'left-bottom'
  | 'right'
  | 'right-top'
  | 'right-bottom'

// 通用触发类型
export type triggerType = 'hover' | 'click' | 'context-menu'

// 转换 width/height/left/... 样式值
export const turnValue = (val: string | number | undefined) => {
  if (!val) {
    return
  }
  if (_.isNumber(val as number)) {
    return `${val}px`
  }
  return val
}

// const _stateValue = ref(defaultValue)
// const innerValue = computed(() => value ?? _stateValue.value)

/**
 * 渲染 default slot，获取子组件 VNode。处理多种子组件创建场景
 * @returns {function(childComponentName: string, slots: Slots): VNode[]}
 * @param childComponentName
 * @param slots
 * @example const getChildByName = useChildComponentSlots()
 * @example getChildComponentByName('xxItem')
 */
export const useChildComponentSlots = () => {
  const instance = getCurrentInstance()
  return (childComponentName: string, slots?: Slots): VNode[] => {
    if (!slots) {
      slots = instance?.slots
    }
    const content = slots?.default?.() || []

    return (
      content
        .map((item: VNode) => {
          if (item.children && Array.isArray(item.children))
            return item.children
          return item
        })
        .flat()
        // @ts-ignore
        .filter((item: VNode) =>
          (item.type as Component).name?.endsWith(childComponentName)
        ) as VNode[]
    )
  }
}
