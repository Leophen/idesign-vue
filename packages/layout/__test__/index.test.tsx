import { Layout, LayoutAside, LayoutContent, LayoutFooter, LayoutHeader } from '../index';
import { render } from '../../common'

describe('Layout 组件测试', () => {
  it('create', () => {
    const wrapper = render(
      <Layout>
        <LayoutHeader>Header</LayoutHeader>
        <Layout>
          <LayoutAside>Aside</LayoutAside>
          <Layout>
            <LayoutContent>Content</LayoutContent>
            <LayoutFooter>Copyright @ 2019-2021 iDesign. All Rights Reserved</LayoutFooter>
          </Layout>
        </Layout>
      </Layout>
    );
    expect(wrapper.element).toMatchSnapshot();
  });

  it('width', () => {
    const wrapper = render(
      <Layout>
        <LayoutHeader>Header</LayoutHeader>
        <Layout>
          <LayoutAside width={100}>Aside</LayoutAside>
          <Layout>
            <LayoutContent>Content</LayoutContent>
            <LayoutFooter>Copyright @ 2019-2021 iDesign. All Rights Reserved</LayoutFooter>
          </Layout>
        </Layout>
      </Layout>
    );
    expect(wrapper.find('.i-layout--aside').attributes('style')).toMatch('width: 100px;');
  })
});

