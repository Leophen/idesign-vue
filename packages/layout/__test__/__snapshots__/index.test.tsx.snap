// Vitest Snapshot v1

exports[`Layout 组件测试 > create 1`] = `
<div
  class="i-layout"
>
  
  <header
    class="i-layout--header"
  >
    
    Header
    
  </header>
  <div
    class="i-layout"
  >
    
    <aside
      class="i-layout--aside"
      style="width: 240px; max-width: 240px; min-width: 240px; flex: 0 0 240px;"
    >
      
      Aside
      
    </aside>
    <div
      class="i-layout"
    >
      
      <main
        class="i-layout--content"
      >
        
        Content
        
      </main>
      <footer
        class="i-layout--footer"
      >
        
        Copyright @ 2019-2021 iDesign. All Rights Reserved
        
      </footer>
      
    </div>
    
  </div>
  
</div>
`;
